### DISCUSSION TITLE
Enum validation with runtime data (VineJS)

### DISCUSSION THREAD

**Discussion by Barbapapazes** [adonisjs/core#4505] [Help]
Hello,

My database contains an array of strings: ['JavaScript', 'TypeScript', 'Python'].

I use these data to create dynamic form where strings are the value of an HTML select.

I would use Vine to validate not only that the received data from this dynamic form is a string but also that data is in the database. I was thinking about something like with `unique`, https://docs.adonisjs.com/guides/validation#passing-metadata-to-validators,  but it does not seems to exists for `enum`.

How could I achieve that? Thanks!

**Comment by Barbapapazes** [comment]
I need to try this https://vinejs.dev/docs/types/enum#defer-computing-enum-options with a field.meta

**Reply by Barbapapazes** [reply]
Could be useful to be able to do the same for the `maxLength` of array

**Comment by thetutlage** [comment]
Yes, you can either use `enum` or the [`in`](https://vinejs.dev/docs/types/string#in) rule. Both support returning runtime values.

Not sure what you mean by `"do the same for the maxLength of array"`. Can you share an example of exactly what you are trying to do?

**Reply by Barbapapazes** [reply]
For example, I have a form where a user can create dynamic form. This user can create a multi-select and indicate how many answers (max) a user can select. This max value is only known at runtime since it comes from the database.

I was trying to use `maxLength` with a custom meta input like with `enum`.

Actually, it's not possible. To solve this, I create a custom `dynamic_max_length_rule`: 

```ts 
import vine from '@vinejs/vine'
import { FieldContext } from '@vinejs/vine/types'

type Options = {}

/**
 * Validates the length of an array based on the field's meta `maxLength` (runtime).
 */
function dynamicMaxLength(value: unknown, options: Options, field: FieldContext) {
  if (!field.isValid) {
    return
  }

  if ((value as unknown[]).length > field.meta.maxLength) {
    field.report(
      `The {{ field }} must not contain more than ${field.meta.maxLength} items`,
      'dynamicMaxLength',
      field,
      options
    )
  }
}

export const dynamicMaxLengthRule = vine.createRule(dynamicMaxLength)
```

This works fine ans I was asking if this is something that could be build-in Vine or maybe a little of documentation explaining this?
